src/celeritas/io/RootStepWriter.cc

```diff
@@ -0,0 +1,188 @@
+
+    // Loop over thread ids and fill TTree
+    for (const auto tid : range(ThreadId{steps.size()}))
+    {
+        if (!steps.track_id[tid])
+        {
+            // Track id not found; skip inactive track slot
+            continue;
+        }
+
+        // Track id is always set
+        tstep_.track_id = steps.track_id[tid].unchecked_get();
+
+        RSW_STORE(event_id, .get());
+        RSW_STORE(action_id, .get());
+        RSW_STORE(energy_deposition, .value());
+        RSW_STORE(step_length, /* no getter */);
+        RSW_STORE(track_step_count, /* no getter */);
+        RSW_STORE_PARTICLE(particle);
+
+        for (const auto i : range(StepPoint::size_))
+        {
+            RSW_STORE(points[i].volume_id, .get());
+            RSW_STORE(points[i].energy, .value());
+            RSW_STORE(points[i].time, /* no getter */);
+            RSW_STORE(points[i].dir, /* no getter */);
+            RSW_STORE(points[i].pos, /* no getter */);
+        }
+
+        tstep_tree_->Fill();
+    }
+
+#undef RSW_STORE
+#undef RSW_STORE_PARTICLE
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Create steps tree. In order to have the option to individually select any
+ * member of `StepStateData` (StepData.hh) to be stored into the ROOT file
+ * *and* not need any dictionary for ROOT I/O, we cannot store an MC truth step
+ * object. Therefore, the data is flattened so that each member of `TStepData`
+ * is an individual branch that stores primitive types and is created based on
+ * the `StepSelection` booleans.
+ *
+ * A macro is usedT to simplify the process of moving from Collection to a ROOT
+ * branch.
+ */
+void RootStepWriter::make_tree()
+{
+#define RSW_CREATE_BRANCH(ATTR, BRANCH_NAME)                      \
+    do                                                            \
+    {                                                             \
+        if (this->selection_.ATTR)                                \
+        {                                                         \
+            this->tstep_tree_->Branch(BRANCH_NAME, &tstep_.ATTR); \
+        }                                                         \
+    } while (0)
+
+    tstep_tree_.reset(new TTree("steps", "steps"));
+
+    tstep_tree_->Branch("track_id", &tstep_.track_id); // Always on
+    RSW_CREATE_BRANCH(event_id, "event_id");
+    RSW_CREATE_BRANCH(track_step_count, "track_step_count");
+    RSW_CREATE_BRANCH(action_id, "action_id");
+    RSW_CREATE_BRANCH(step_length, "step_length");
+    RSW_CREATE_BRANCH(particle, "particle");
+    // Pre-step
+    RSW_CREATE_BRANCH(points[StepPoint::pre].volume_id, "point_pre_volume_id");
+    RSW_CREATE_BRANCH(points[StepPoint::pre].dir, "point_pre_dir");
+    RSW_CREATE_BRANCH(points[StepPoint::pre].pos, "point_pre_pos");
+    RSW_CREATE_BRANCH(points[StepPoint::pre].energy, "point_pre_energy");
+    RSW_CREATE_BRANCH(points[StepPoint::pre].time, "point_pre_time");
```

Comment:
Do you think it might be better to drop the `point_` prefix?

You could also loop over the pre/post values by:
```c++
        const EnumArray<StepPoint, std::string> prefixes{{"pre_", "post_"}};
        for (auto sp : range(StepPoint::size_))
        {
            RSW_CREATE_BRANCH(points[sp].time, prefixes[sp] + "time");
// ...
        }
```

Reply:
The `prefixes[sp] + "time"` will not expand correctly into the macro. So I'd have to concatenate 2 strings and provide a const char* to the macro (`Branch(...)` only accepts const char*), which seems to make it more convoluted than we'd like.

Reply:
Ah, I get what you're saying. The complexity tradeoff is fine for now.